executor/admin.go

```diff
@@ -163,6 +168,80 @@ func (e *CheckIndexRangeExec) Close() error {
 	return nil
 }
 
+// RestoreTableExec represents a recover table executor.
+// It is built from "admin restore table by job" statement,
+// is used to recover the table that deleted by mistake.
+type RestoreTableExec struct {
+	baseExecutor
+	jobID int64
+}
+
+// Open implements the Executor Open interface.
+func (e *RestoreTableExec) Open(ctx context.Context) error {
+	if err := e.baseExecutor.Open(ctx); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+// Next implements the Executor Open interface.
+func (e *RestoreTableExec) Next(ctx context.Context, chk *chunk.Chunk) error {
+	txn, err := e.ctx.Txn(true)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	t := meta.NewMeta(txn)
+	job, err := t.GetHistoryDDLJob(e.jobID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if job == nil {
+		return admin.ErrDDLJobNotFound.GenWithStackByArgs(e.jobID)
+	}
+	if job.Type != model.ActionDropTable {
+		return errors.Errorf("Job %v type is %v, not drop table", job.ID, job.Type)
+	}
+
+	// check gc safe point
+	err = gcutil.ValidateSnapshot(e.ctx, job.StartTS)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	dom := domain.GetDomain(e.ctx)
+	// Get the snapshot infoSchema before drop table.
+	snapInfo, err := dom.GetSnapshotInfoSchema(job.StartTS)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	// Get table meta from snapshot infoSchema.
+	table, ok := snapInfo.TableByID(job.TableID)
+	if !ok {
+		return infoschema.ErrTableNotExists.GenWithStackByArgs(
+			fmt.Sprintf("(Schema ID %d)", job.SchemaID),
+			fmt.Sprintf("(Table ID %d)", job.TableID),
+		)
+	}
+	// Get table original autoID before table drop.
+	m, err := dom.GetSnapshotMeta(job.StartTS)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	autoID, err := m.GetAutoTableID(job.SchemaID, job.TableID)
+	if err != nil {
+		return errors.Errorf("recover table_id: %d, get original autoID from snapshot meta err: %s", job.TableID, err.Error())
+	}
+	// check gc enable use to decide whether enable gc after restore table.
+	gcEnable, err := gcutil.CheckGCEnable(e.ctx)
```

Comment:
We can CheckGCEnable here because the executing is concurrency. Please check the status always in the DDL owner.